Implement automatic differentiation engine

helloheyi opened the pull request:
Summary
This PR introduces an automatic differentiation engine using object-oriented C++. It supports both forward and backward a computational graph of operations e.g.+,/,-, etc.).

Key Components
  - Node classes: AddNode, SubNode, MulNode and etc 
  - Tape class: store and track nodes forward/backward passes
  - Examples in main.cpp

Features
  - Forward pass computes function value
  - Back-propagation computes gradients 
  - Finite difference function included for gradient and function verification
  - Edge case handling

Testing
  - Compared gradients with finite difference results.
  - Evaluated multiple test cases including:
  - Single-variable expressions
  - Multi-variable functions
  - Edge cases (e.g. x = -1, 0, large values)
  - Output validated via std::cout in main.cpp.

Next Steps
  - Add more complex node types (e.g., `SinNode`, `ExpNode`)
  - Handle higher-order or more complex multivariable functions

helloheyi committed:
Update REMADME

helloheyi committed:
Update README

helloheyi committed:
Update README.md

helloheyi committed:
Update README.md

helloheyi committed:
Update README.md

helloheyi committed:
Redesign a Node class for runtime polymorphism.

Redesign  framework to Improve: 
1.  Implement a Node class for runtime polymorphism to enhance scalability.
2. Add clear and detailed comments for better readability.
3.  Incorporate edge case handling.
4. Create testing for one variable function, two variable function, cut edge scenarios

helloheyi committed:
Replace pointer to shared smart pointers 

Enhancement: Use smart pointers for efficient memory management

helloheyi committed:
Upload results from enhancement version

helloheyi committed:
Improve: multiline block comment style to classes

Applied a multi-line block comment style across all classes to enhance code readability and maintainability. 
The comments now provide clear descriptions of class functionality, parameters, and method behavior, following a good practices for documentation.

christophercylai commented:
Hi, this is Chris. I am one of the interviewers. I am going to give comments and recommendations.
- I appreciate the comments, very helpful - despite a bit long
- The design of this project is over complicated: the expectation is minimum coding to satisfy the task at hand
- Please remove the taping/forward/backward
- Instead of making Node a parent class of Add (and others), can Add just be a class method of Node?
- Bonus: put everything into the main.cpp
- Bonus: use gcc to compile (https://code.visualstudio.com/docs/cpp/config-mingw), and give me the compilation command

christophercylai commented on main.cpp:
```diff
@@ -0,0 +1,171 @@
+}
+
+
+int main() {
```
- main function should take in the x_value from the command line (example: fte.exe 5)
- main function should write the results to a text file, without printing out the results (do print out the name of the text file that the results are written to)

helloheyi replied:
Thanks for the detailed feedback, Chris!


Updated `main.cpp` to:
- Accept `x` as a command-line argument (e.g., `./autoDiff 2`)
- Write the results (f(x), df/dx) to `results.txt` instead of printing them to stdout
- Print only the file name as confirmation, as requested

Let me know if there's anything else you'd like adjusted.

christophercylai commented on main.cpp:
```diff
@@ -0,0 +1,171 @@
+   - f(x_0) (Function value)
+   - f'(x_0) (Gradient using reverse-mode autodiff)
+*/
+auto runGraphTest1(double x_0) {
```
move this into main()
also, there's no need to create so many variables - the node class should understand operators such as '+' and '-'

helloheyi replied:
- Updated main.cpp to build expression trees directly
- Each Node handles its own forward and backward pass logic

christophercylai commented on main.cpp:
```diff
@@ -0,0 +1,171 @@
+   - df/dx (Gradient w.r.t. x)
+   - df/dy (Gradient w.r.t. y)
+*/
+auto runGraphTest2(double x_0, double y_0) {
```
move this into main()

helloheyi replied:
Updated main.cpp to build expression trees directly
Each Node handles its own forward and backward pass logic

helloheyi committed:
Refactor autodiff engine: removed Tape, structured operations as Node subclasses

- Deleted Tape.cpp and Tape.hpp (no longer used)
- Refactored design to use polymorphic Node subclasses (AddNode, MulNode, etc.)
- Each Node handles its own forward and backward pass logic
- Updated main.cpp to build expression trees directly
- Supports one-variable and two-variable functions
- CLI-ready (e.g., ./autoDiff 5), with results saved to results.txt
- Updated README.md to reflect the new object-oriented design
- Included autoDiff executable for review/testing

helloheyi committed:
Redesign operation by class-method-based Node based on PR feedback

- Removed the taping/forward/backward logic
- Instead of making operations as derived classes from a base class,
  changed Add to be a class method of Node
- Demonstrates minimal design with operations implemented directly within Node
- Includes main.cpp, node.cpp, node.h, and results.txt for testing
- Uses gcc to compile with:

  g++ -std=c++14 main.cpp node.cpp -o class_methods

  Then run with:
  ./class_methods 2

helloheyi committed:
Redesign operation by class-method-based Node based on PR feedback - Removed the taping/forward/backward logic
- Instead of making operations as derived classes from a base class,
  changed Add to be a class method of Node
- Demonstrates minimal design with operations implemented directly within Node
- Includes main.cpp, node.cpp, node.h, and results.txt for testing
- Uses gcc to compile with:
  g++ -std=c++14 main.cpp node.cpp -o class_methods
  Then run with: ./class_methods 2

helloheyi commented:
Hi Chris, thanks again for the thoughtful and direct feedback — much appreciated! 

I've incorporated your suggestions into two separate versions.

In Version 1 (the previous design)
  - I removed the Tape structure and recursive forward/backward infrastructure while keeping the object-oriented approach. 
  - Each operation (e.g., AddNode, PowNode) still inherits from a common Node base class, and gradients are computed via 
  - polymorphic method calls. While this version is more extensible, I recognize it introduces some extra complexity. 
  - Updated main.cpp to build expression trees
  -  Updated README

This version can be compiled g++ -std=c++14     main.cpp     nodes/ConstantNode.cpp     nodes/VariableNode.cpp     nodes/AddNode.cpp     nodes/SubNode.cpp     nodes/MulNode.cpp     nodes/DivNode.cpp     nodes/PowNode.cpp     nodes/LogNode.cpp     -o autoDiff
./autoDiff 2

In Version 2 (class_methods_of_Node/), 
I implemented a minimal design where each operation (Add, Sub, etc.) is defined directly as a class method of Node rather than using derived subclasses. 
  - This structure is more compact and clean, with the logic separated into main.cpp, node.cpp, and node.h. It aligns more 
     closely with your minimal code philosophy.
  - Updated main.cpp to build expression trees

This version can be compiled with g++ -std=c++14 main.cpp node.cpp -o class_methods, and run with ./class_methods 2

These executable files are uploaded as well. Let me know if you'd prefer an even more condensed version fully contained in main.cpp — happy to streamline further if needed!

christophercylai commented:
Thanks! I am going to try this.
